### Dataset 1: Austin

In [63]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression, Ridge, Lasso, ElasticNet
from sklearn.preprocessing import StandardScaler, PolynomialFeatures
from sklearn.metrics import r2_score

In [64]:
# Load dataset
austin_df = pd.read_csv("/Users/sohanaziz/Documents/AI for leaders sem 2/Datasets/cleaned datasets/week 2/cleaned_austin_data.csv")

austin_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15171 entries, 0 to 15170
Data columns (total 44 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   city                            15171 non-null  object 
 1   zipcode                         15171 non-null  int64  
 2   propertyTaxRate                 15171 non-null  float64
 3   garageSpaces                    15171 non-null  int64  
 4   hasAssociation                  15171 non-null  bool   
 5   hasCooling                      15171 non-null  bool   
 6   hasHeating                      15171 non-null  bool   
 7   hasSpa                          15171 non-null  bool   
 8   hasView                         15171 non-null  bool   
 9   parkingSpaces                   15171 non-null  int64  
 10  yearBuilt                       15171 non-null  int64  
 11  latestPrice                     15171 non-null  float64
 12  numPriceChanges                 

In [65]:
austin_df.head()

,city,zipcode,propertyTaxRate,garageSpaces,hasAssociation,hasCooling,hasHeating,hasSpa,hasView,parkingSpaces,...,numOfStories,homeType_Condo,homeType_Mobile / Manufactured,homeType_MultiFamily,homeType_Multiple Occupancy,homeType_Other,homeType_Residential,homeType_Single Family,homeType_Townhouse,homeType_Vacant Land
0,pflugerville,78660,1.98,2,True,True,True,False,False,2,...,2,0,0,0,0,0,0,1,0,0
1,pflugerville,78660,1.98,2,True,True,True,False,False,2,...,1,0,0,0,0,0,0,1,0,0
2,pflugerville,78660,1.98,0,True,True,True,False,False,0,...,1,0,0,0,0,0,0,1,0,0
3,pflugerville,78660,1.98,2,True,True,True,False,False,2,...,1,0,0,0,0,0,0,1,0,0
4,pflugerville,78660,1.98,0,True,True,True,False,False,0,...,2,0,0,0,0,0,0,1,0,0


In [66]:
# Select numeric features and target variable
features = ['livingAreaSqFt', 'numOfBathrooms', 'numOfBedrooms', 'lotSizeSqFt', 'avgSchoolRating', 'yearBuilt']
X = austin_df[features]
y = austin_df['latestPrice']

**original features**

In [67]:
# Split the data into train and test sets (original features)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Define models to evaluate
models = {
    "Linear": LinearRegression(),
    "Ridge": Ridge(alpha=1.0),
    "Lasso": Lasso(alpha=0.1),
    "ElasticNet": ElasticNet(alpha=0.1, l1_ratio=0.5)
}

In [68]:
# Scale original features
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)


In [69]:
# Train and evaluate models on original features
for name, model in models.items():
    model.fit(X_train_scaled, y_train)
    y_pred = model.predict(X_test_scaled)
    r2 = r2_score(y_test, y_pred)
    print(f"{name} Regression R²: {r2:.4f}")

Linear Regression R²: 0.3252
Ridge Regression R²: 0.3252
Lasso Regression R²: 0.3252
ElasticNet Regression R²: 0.3213


**polynomial features**

In [70]:
# Create polynomial and interaction features (degree=2)
poly = PolynomialFeatures(degree=2, include_bias=False, interaction_only=False)
X_poly = poly.fit_transform(X)

# Split polynomial features data
X_train_poly, X_test_poly, y_train_poly, y_test_poly = train_test_split(X_poly, y, test_size=0.2, random_state=42)


In [71]:
# Scale polynomial features
scaler_poly = StandardScaler()
X_train_poly_scaled = scaler_poly.fit_transform(X_train_poly)
X_test_poly_scaled = scaler_poly.transform(X_test_poly)

In [72]:
# Train and evaluate models on polynomial features
for name, model in models.items():
    model.fit(X_train_poly_scaled, y_train_poly)
    y_pred_poly = model.predict(X_test_poly_scaled)
    r2_poly = r2_score(y_test_poly, y_pred_poly)
    print(f"{name} Regression with Poly Features R²: {r2_poly:.4f}")

Linear Regression with Poly Features R²: 0.3795
Ridge Regression with Poly Features R²: 0.4241


/Users/sohanaziz/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_coordinate_descent.py:530: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 573025872485448.4, tolerance: 239726144730.80865
  model = cd_fast.enet_coordinate_descent(


Lasso Regression with Poly Features R²: 0.4213
ElasticNet Regression with Poly Features R²: 0.4166


### Dataset 2: Airbnb 

In [73]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, PolynomialFeatures
from sklearn.linear_model import LinearRegression, Ridge, Lasso, ElasticNet
from sklearn.metrics import r2_score

# Load cleaned dataset
df = pd.read_csv("/Users/sohanaziz/Documents/AI for leaders sem 2/Datasets/cleaned datasets/week 2/listings_cleaned.csv")

# Select numeric features for regression
features = [
    'latitude', 'longitude', 'minimum_nights', 'number_of_reviews', 'reviews_per_month',
    'calculated_host_listings_count', 'availability_365', 'bedrooms', 'beds', 'bathrooms',
    'num_amenities', 'host_tenure_years',
    'amenity_wifi', 'amenity_kitchen', 'amenity_air_conditioning', 'amenity_heating', 'amenity_tv',
    'host_is_superhost', 'host_has_profile_pic', 'host_identity_verified', 'instant_bookable'
]

X = df[features]
y = df['price']

# Train-test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Scale features
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Define models
models = {
    "Linear": LinearRegression(),
    "Ridge": Ridge(alpha=1.0),
    "Lasso": Lasso(alpha=0.1),
    "ElasticNet": ElasticNet(alpha=0.1, l1_ratio=0.5)
}

print("---- Original Features ----")
for name, model in models.items():
    model.fit(X_train_scaled, y_train)
    y_pred = model.predict(X_test_scaled)
    r2 = r2_score(y_test, y_pred)
    print(f"{name} Regression R²: {r2:.4f}")

# Polynomial features (degree 2)
poly = PolynomialFeatures(degree=2, include_bias=False)
X_poly = poly.fit_transform(X)

# Train-test split with polynomial features
X_train_poly, X_test_poly, y_train_poly, y_test_poly = train_test_split(X_poly, y, test_size=0.2, random_state=42)

# Scale polynomial features
scaler_poly = StandardScaler()
X_train_poly_scaled = scaler_poly.fit_transform(X_train_poly)
X_test_poly_scaled = scaler_poly.transform(X_test_poly)

print("\n---- Polynomial Features (Degree=2) ----")
for name, model in models.items():
    model.fit(X_train_poly_scaled, y_train_poly)
    y_pred_poly = model.predict(X_test_poly_scaled)
    r2_poly = r2_score(y_test_poly, y_pred_poly)
    print(f"{name} Regression with Poly Features R²: {r2_poly:.4f}")

---- Original Features ----
Linear Regression R²: 0.1360
Ridge Regression R²: 0.1360
Lasso Regression R²: 0.1363
ElasticNet Regression R²: 0.1398

---- Polynomial Features (Degree=2) ----
Linear Regression with Poly Features R²: -0.4222
Ridge Regression with Poly Features R²: -0.0877


/Users/sohanaziz/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_coordinate_descent.py:530: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 688255682.7647748, tolerance: 419314.9319261829
  model = cd_fast.enet_coordinate_descent(


Lasso Regression with Poly Features R²: -0.0483
ElasticNet Regression with Poly Features R²: 0.1976


/Users/sohanaziz/opt/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_coordinate_descent.py:530: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 511387617.6888664, tolerance: 419314.9319261829
  model = cd_fast.enet_coordinate_descent(


**Austin Housing Dataset**

My initial model using just the original features explained about 32% of the variation in housing prices, which suggests a moderate linear relationship between the input features (like square footage, number of bathrooms, and school rating) and price. This is a reasonable outcome for real estate data, where prices are influenced by many structured and unstructured factors.

When I added polynomial features, such as squared terms and interactions between variables, the model was able to capture nonlinear patterns in the data. As a result, the R² score increased to around 38% for standard linear regression, showing that more complex relationships were now being modeled.

I also applied regularized regression models like Ridge, Lasso, and ElasticNet, which help prevent overfitting by penalizing overly complex models. Among these, Ridge regression with polynomial features performed best, improving the R² to about 42%. This indicates that the model could generalize well and provide better predictive performance while managing multicollinearity and overfitting risks.